### PR TITLE
nixos/nvidia: include systemd power management

### PR DESCRIPTION
**Pull request by eadwu**
###### Motivation for this change
https://github.com/NixOS/nixpkgs/issues/73494

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: nixos/modules/hardware/video/nvidia.nix [hunk 1]
```diff
@@ -180,6 +189,25 @@ in
...
+        description = "NVIDIA system ${type} actions";
+        before = [ "systemd-${type}.service" ];
+
+        path = with pkgs; [ coreutils nvidia_x11.bin ];
+        serviceConfig = {
+          Type = "oneshot";
+        };
+
+        script = ''
+          logger -t ${type} -s "nvidia-${type}.service"
```
Comment (flokli) [inline nixos/modules/hardware/video/nvidia.nix]: systemd should automatically log when it's enabling services. You should be able to just put the `nvidia-sleep.sh` call into `serviceConfig.ExecStart`.

File: nixos/modules/hardware/video/nvidia.nix [hunk 2]
```diff
@@ -180,6 +189,25 @@ in
     environment.systemPackages = [ nvidia_x11.bin nvidia_x11.settings ]
       ++ lib.filter (p: p != null) [ nvidia_x11.persistenced ];
 
+    systemd.services = let
+      nvidiaService = type: {
```
Comment (flokli) [inline nixos/modules/hardware/video/nvidia.nix]: I don't really think this should be called type, maybe `sleepState`, like in `systemd-suspend.service`?

File: nixos/modules/hardware/video/nvidia.nix [hunk 3]
```diff
@@ -180,6 +189,25 @@ in
...
+        };
+
+        script = ''
+          logger -t ${type} -s "nvidia-${type}.service"
+          nvidia-sleep.sh "${type}"
+        '';
+
+        requiredBy = [ "systemd-${type}.service" ];
+      };
+    in builtins.listToAttrs (map (t: nameValuePair "nvidia-${t}" (nvidiaService t)) ["hibernate" "resume" "suspend"]);
```
Comment (flokli) [inline nixos/modules/hardware/video/nvidia.nix]: Reading through `man systemd-suspend.service`, I don't really get how this should work with `systemd-resume.service`, and wonder if `hybrid-sleep` and `suspend-then-hibernate` are missing.

Would it be possible to add this to `/usr/lib/systemd/system-sleep/`? I assume nvidia also doesn't support https://www.freedesktop.org/wiki/Software/systemd/inhibit/, do they?

Comment (eadwu) [reply nixos/modules/hardware/video/nvidia.nix]: They have both the script located in `lib/systemd/system-sleep/nvidia` and the service. As for the service for resuming, I messed up the `before` and `requiredBy` attributes.

File: nixos/modules/hardware/video/nvidia.nix [hunk 4]
```diff
@@ -180,6 +189,33 @@ in
...
+
+      nvidiaService = sleepState: (baseNvidiaService sleepState) // {
+        before = [ "systemd-${sleepState}.service" ];
+        requiredBy = [ "systemd-${sleepState}.service" ];
+      };
+
+      services = (builtins.listToAttrs (map (t: nameValuePair "nvidia-${t}" (nvidiaService t)) ["hibernate" "suspend"]))
+        // {
+          nvidia-resume = (baseNvidiaService "resume") // {
+            before = [ "systemd-suspend.service" "systemd-hibernate.service" ];
```
Comment (eadwu) [inline nixos/modules/hardware/video/nvidia.nix]: Should be `after`.

File: nixos/modules/hardware/video/nvidia.nix [hunk 5]
```diff
@@ -39,6 +39,15 @@ in
 
 {
   options = {
+    hardware.nvidia.powerManagement.enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Experimental power management though systemd. For more information, see
```
Comment (mweinelt) [inline nixos/modules/hardware/video/nvidia.nix]: through

**Comment by lovesegfault** [comment]
What's the status on this @eadwu @flokli?

**Comment by eadwu** [comment]
I'm waiting for #66601 to get merged before pushing for this since it will introduce conflicts.

**Comment by Mic92** [comment]
https://github.com/NixOS/nixpkgs/pull/66601 is merged now.

File: nixos/modules/hardware/video/nvidia.nix [hunk 6]
```diff
@@ -52,6 +52,15 @@ in
     ];
 
   options = {
+    hardware.nvidia.powerManagement.enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Experimental power management through systemd. For more information, see
+        the NVIDIA docs, on Chapter 21. Configuring Power Management Support.
```
Comment (lovesegfault) [inline nixos/modules/hardware/video/nvidia.nix:60]: If these docs are publicly available, then a link would be nice.

Comment (eadwu) [reply nixos/modules/hardware/video/nvidia.nix:60]: The docs are within each package, so pinning the doc has the possibility of the docs becoming outdated (I'm not sure if there's a link for the "latest" package).

Then again, I could probably do something along the lines of `https://download.nvidia.com/XFree86/Linux-x86_64/${nvidia_x11.version}/README/powermanagement.html` but not sure if a package evaluation should be done just for the option.

**Review by lovesegfault — APPROVED** [review]
:shipit:

File: nixos/modules/hardware/video/nvidia.nix (hunk 6 above)
Comment (lovesegfault) [reply nixos/modules/hardware/video/nvidia.nix:60]: Fair :)

File: nixos/modules/hardware/video/nvidia.nix [hunk 7]
```diff
@@ -226,31 +235,60 @@ in
...
+    systemd.packages = optional cfg.powerManagement.enable nvidia_x11.out;
+
+    systemd.services = let
+      baseNvidiaService = state: {
+        description = "NVIDIA system ${state} actions";
+
+        path = with pkgs; [ kbd ];
+        serviceConfig = {
+          Type = "oneshot";
+          ExecStart = "${nvidia_x11.out}/bin/nvidia-sleep.sh '${state}'";
```
Comment (nh2) [inline nixos/modules/hardware/video/nvidia.nix:247]: Broken with new driver due to conditional install (silent failure in package install process from this PR) if the script doesn't exist:

https://github.com/NixOS/nixpkgs/issues/127693

**Pull request merged**